server/ocsp.go

```diff
@@ -0,0 +1,377 @@
+
+	return raw, resp, nil
+}
+
+func (oc *OCSPMonitor) run() {
+	s := oc.srv
+	s.mu.Lock()
+	quitCh := s.quitCh
+	s.mu.Unlock()
+
+	defer s.grWG.Done()
+	config := s.getOpts().OCSPConfig
+	shouldShutdown := config != nil && (config.Mode == OCSPModeMust || config.Mode == OCSPModeAlways)
+
+	var nextRun time.Duration
+	_, resp, err := oc.getStatus()
+	if err == nil && resp.Status == ocsp.Good {
+		nextRun = oc.getNextRun()
+		t := resp.NextUpdate.Format(time.RFC3339Nano)
+		s.Noticef(
+			"Found existing OCSP certificate status: good, next update %s, checking again in %s",
+			t, nextRun,
+		)
+	} else if err == nil && shouldShutdown {
+		// If resp.Status is ocsp.Revoked, ocsp.Unknown, or any other value.
+		s.Errorf("Found existing OCSP certificate status: %s", ocspStatusString(resp.Status))
+		s.Shutdown()
+		return
+	}
+
+	for s.Running() {
+		select {
+		case <-time.After(nextRun):
+		case <-quitCh:
+			return
+		}
+		_, resp, err := oc.getRemoteStatus()
+		if err != nil {
+			nextRun = oc.getNextRun()
+			s.Errorf("Bad OCSP status update: %s, trying again in %v", err, nextRun)
+			continue
+		}
+
+		switch n := resp.Status; n {
+		case ocsp.Good:
+			nextRun = oc.getNextRun()
+			t := resp.NextUpdate.Format(time.RFC3339Nano)
+			s.Noticef(
+				"Received OCSP certificate status: good, next update %s, checking again in %s",
+				t, nextRun,
+			)
+			continue
+		default:
+			s.Errorf("Received OCSP certificate status: %s, shutting down", ocspStatusString(n))
+			s.Shutdown()
+			return
+		}
+	}
+}
+
+// NewOCSPMonitor takes a TLS configuration then wraps it with the callbacks set for OCSP verification
+// along with a monitor that will periodically fetch OCSP staples.
+func (srv *Server) NewOCSPMonitor(tc *tls.Config) (*tls.Config, *OCSPMonitor, error) {
+	opts := srv.getOpts()
+	oc := opts.OCSPConfig
+	tcOpts := opts.TLSConfigOpts
+	if oc.Mode == OCSPModeNever {
+		return tc, nil, nil
+	}
+	cert, err := tls.LoadX509KeyPair(tcOpts.CertFile, tcOpts.KeyFile)
+	if err != nil {
+		return nil, nil, err
+	}
+	cert.Leaf, err = x509.ParseCertificate(cert.Certificate[0])
+	if err != nil {
+		return nil, nil, err
+	}
+	if oc.Mode == OCSPModeAuto && !hasOCSPStatusRequest(cert.Leaf) {
+		// "status_request" flag not set. No need to do anything.
+		return tc, nil, nil
+	}
+
+	// TODO: Add OCSP 'responder_cert' option in case CA cert not available.
+	caCert := tcOpts.CaFile
+	issuer, err := getOCSPIssuer(caCert, cert.Certificate)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Callbacks below will be in charge of returning the certificate instead,
+	// so this has to be nil.
+	tc.Certificates = nil
+
+	mon := &OCSPMonitor{
+		srv: srv,
+		hc:  &http.Client{Timeout: 30 * time.Second},
+		// FIXME: Make reloadable aware, otherwise a change in certificates
+		//  could make the staples stale.
+		Leaf:   cert.Leaf,
+		Issuer: issuer,
+	}
+
+	// Get the certificate status from the memory, disk, then remote OCSP responder.
+	_, resp, err := mon.getStatus()
```

Comment:
This is missing handling for the staple in the disk cache being stale at start and needing to be replaced immediately.

Reply:
Currently it always fetches from the remote OCSP server, we will address that in the next PR that includes caching to disk.